### PR TITLE
TASK 5 PR

### PR DESCRIPTION
**Pull request by ZaniInc**

### CODE REVIEW

File: contracts/MyNFT.sol [hunk 1]
```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.7;
+
+// Import this file to use console.log
+import "hardhat/console.sol";
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: unused import

File: contracts/MyNFT.sol [hunk 2]
```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: default license is MIT/ISC

Comment (BhHaipls) [reply contracts/MyNFT.sol]: for sc

File: contracts/MyNFT.sol [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+
+/**
+ * @title Nft
+ * @author Pavel Zanozin
+ * @notice This SC allows users buy NFT and register
+ * ownership on blockchain
+ */
+contract MyNFT is ERC721, Ownable {
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: interface for your NFT is miss

File: contracts/MyNFT.sol [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    event SetNFTPrice(uint256 price);
+
+    /**
+     * @dev event logs info base URI , when owner call
+     * 'setBaseURI'
+     *
+     * @param baseURILink_ - contain base URI link for all
+     * NFT's
+     */
+    event SetBaseURI(string baseURILink_);
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: param for event should be without _ in end of name

File: contracts/MyNFT.sol [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    /**
+     * @dev event logs info about who's buy nft and
+     * how many ether's transfer to SC
+     *
+     * @param buyer - who's buy
+     * @param ethers - how many ether's transfer
+     */
+    event Buy(address buyer, uint256 ethers);
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: should add id of nft which was mint for buyer

File: contracts/MyNFT.sol [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+     * @dev allows user buy nft's for the price which set
+     * by the owner . Don't take any params
+     *
+     * @notice can be by call by any one . revert transaction if
+     * all nft's was minted or paymant value too low
+     */
+    function buy() external payable {
+        require(msg.value == nftPrice, "Error : payment value too low");
+        require(
+            tokenCountId._value < tokenSupply,
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: use current() method from Counter instead of _value.

File: contracts/MyNFT.sol [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            "Error : all NFT's was sold"
+        );
+        tokenCountId.increment();
+        _safeMint(msg.sender, tokenCountId.current());
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: in this case, nft will start from 1 id. should start from zero

File: contracts/MyNFT.sol [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        return baseURILink;
+    }
+
+    /**
+     * @dev allows owner set base URI for all NFT's
+     *
+     * @param baseURI_ - input base URI
+     * @notice can be call only by owner , 'baseURI_' can't be 0
+     */
+    function setBaseURI(string memory baseURI_) public onlyOwner {
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: should be internal

File: contracts/MyNFT.sol [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        view
+        virtual
+        override
+        returns (string memory)
+    {
+        _requireMinted(tokenId);
+
+        string memory baseURI = _baseURI();
+        return
+            bytes(baseURI).length > 0
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: it's not need check, because baseURI always will have length more then 0

File: contracts/MyNFT.sol (hunk 8 above)
Comment (BhHaipls) [reply contracts/MyNFT.sol]: and call only from constructor

File: scripts/deploy.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+// We require the Hardhat Runtime Environment explicitly here. This is optional
+// but useful for running the script in a standalone fashion through `node <script>`.
+//
+// You can also run a script with `npx hardhat run <script>`. If you do that, Hardhat
+// will compile your contracts, add the Hardhat Runtime Environment's members to the
+// global scope, and execute the script.
+const hre = require("hardhat");
+
```
Comment (BhHaipls) [inline scripts/deploy.js:3]: it's script file should be change/delete it's not from this task

**Review by BhHaipls — CHANGES_REQUESTED** [review]

File: contracts/MyNFT.sol [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "./interfaces/IMyNFT.sol";
+
+/**
+ * @title Nft
+ * @author Pavel Zanozin
+ * @notice This SC allows users buy NFT and register
+ * ownership on blockchain
+ */
+contract MyNFT is ERC721, Ownable, IMyNFT {
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: interfaces have priority, in order in extends list

File: contracts/MyNFT.sol [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+     * @dev allows user buy nft's for the price which set
+     * by the owner . Don't take any params
+     *
+     * @notice can be by call by any one . revert transaction if
+     * all nft's was minted or paymant value too low
+     */
+    function buy() external payable override {
+        require(msg.value == nftPrice, "Error : payment value too low");
+        require(
+            tokenCountId.current() <= tokenSupply,
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: for gas optimization create variable for tokenCountId.current() for interact with memory and not interact with storage every time

File: contracts/MyNFT.sol [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+        return baseURILink;
+    }
+
+    /**
+     * @dev allows owner set base URI for all NFT's
+     *
+     * @param baseURI_ - input base URI
+     * @notice can be call only by owner , 'baseURI_' can't be 0
+     */
+    function setBaseURI(string memory baseURI_) internal onlyOwner {
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: internal function name should start with "_"

File: contracts/MyNFT.sol [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        public
+        view
+        virtual
+        override
+        returns (string memory)
+    {
+        _requireMinted(tokenId);
+
+        string memory baseURI = _baseURI();
+        return string(abi.encodePacked(baseURI, tokenId.toString(), ".json"));
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: just use direct call _baseURI() instead create memory variable

File: contracts/interfaces/IMyNFT.sol [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+     * NFT's
+     */
+    event SetBaseURI(string URILink);
+
+    /**
+     * @dev event logs info about who's buy nft and
+     * how many ether's transfer to SC
+     *
+     * @param buyer - who's buy
+     * @param ethers - how many ether's transfer
```
Comment (BhHaipls) [inline contracts/interfaces/IMyNFT.sol:34]: tokenId is miss in param list in documentation

File: contracts/MyNFT.sol [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    function buy() external payable override {
+        require(msg.value == nftPrice, "Error : payment value too low");
+        require(
+            tokenCountId.current() <= tokenSupply,
+            "Error : all NFT's was sold"
+        );
+        _safeMint(msg.sender, tokenCountId.current());
+        emit Buy(msg.sender, tokenCountId.current(), msg.value);
+        tokenCountId.increment();
```
Comment (BhHaipls) [inline contracts/MyNFT.sol:70]: increment operation should be before emit event
follow next operation ordering:
1) require
2) calculation
4) change storage
5) transfers
6) event

it's by default, but you can change ordering if needed but not in this case

File: test/MyNFT.test.js [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+contract("MyNFT", async ([owner, acc1, acc2, acc3, acc4, acc5, acc6]) => {
+
+  let baseURI = "https://gateway.pinata.cloud/ipfs/QmcUH66aaS6iCLJdDCas845xkzcRYgq8NNwfs3FMir8wTs/";
+
+  before(async () => {
+    instanceNFT = await MyNFT.new(ether('0.001'), baseURI);
+  });
+
+  describe("Initialize contract - false", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: add tests for you check the parameters initialize is corectly

Comment (BhHaipls) [reply test/MyNFT.test.js]: price, baseURI, maxTotalSupply etc

File: contracts/MyNFT.sol [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * @notice can be changed by owner.
+     */
+    uint256 public nftPrice;
+
+    /**
+     * @dev contain max tokens supply
+     *
+     */
+    uint256 public tokenSupply = 5;
```
Comment (BhHaipls) [inline contracts/MyNFT.sol]: you never change this variable, should be constant

File: contracts/MyNFT.sol [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+    /**
+     * @dev contain baseURI.
+     *
+     * @notice can be changed by owner.
+     */
+    string public baseURILink;
```
Comment (BhHaipls) [inline contracts/MyNFT.sol:46]: should be immutable type

File: test/MyNFT.test.js [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+      await expectRevert(MyNFT.new(ether('0'), baseURI), "Error : nft price can't be equal 0");
+    });
+    it("Error : not valid URI", async () => {
+      await expectRevert(MyNFT.new(ether('1'), ""), "Error : not valid URI");
+    });
+  });
+
+  describe("setNftPrice function", async () => {
+
+    describe("False", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: false it's not corectly name for describe with fail tests
rename to: "Should fail if..."

File: test/MyNFT.test.js [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+  describe("Initialize contract - false", async () => {
+    it("Error : nft price can't be equal 0", async () => {
+      await expectRevert(MyNFT.new(ether('0'), baseURI), "Error : nft price can't be equal 0");
+    });
+    it("Error : not valid URI", async () => {
+      await expectRevert(MyNFT.new(ether('1'), ""), "Error : not valid URI");
+    });
+  });
+
+  describe("setNftPrice function", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: just setNftPrice without function suffix in name

File: test/MyNFT.test.js [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+
+    describe("False", async () => {
+      it("Ownable: caller is not the owner", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: call is not owner

File: test/MyNFT.test.js [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+        let currentPrice = await instanceNFT.nftPrice();
+        expect(currentPrice).to.be.bignumber.equal(ether('0.001'));
+        await expectRevert(instanceNFT.setNftPrice(ether('1'), { from: acc6 }), "Ownable: caller is not the owner");
+        let currentPriceAfter = await instanceNFT.nftPrice();
+        expect(currentPriceAfter).to.be.bignumber.equal(ether('0.001'));
+      });
+    });
+
+    describe("Done", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: Success...

File: test/MyNFT.test.js [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+      it("change price by owner", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: change price by owner and emit events

File: test/MyNFT.test.js [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+
+    describe("Done", async () => {
+      it("buy by owner", async () => {
+        let balanceBefore = await instanceNFT.balanceOf(owner);
+        expect(balanceBefore).to.be.bignumber.equal(new BN(0));
+        tx = await instanceNFT.buy({ value: ether('1') });
+        let balanceAfter = await instanceNFT.balanceOf(owner);
+        expect(balanceAfter).to.be.bignumber.equal(new BN(1));
+        let tokenURI = await instanceNFT.tokenURI(new BN(0));
+        expect(tokenURI).to.be.equal("https://gateway.pinata.cloud/ipfs/QmcUH66aaS6iCLJdDCas845xkzcRYgq8NNwfs3FMir8wTs/0.json");
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: you should check tokenURI in another specific tests

File: test/MyNFT.test.js [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+        let balanceAfterNFT = await instanceNFT.balanceOf(acc5);
+        expect(balanceAfterNFT).to.be.bignumber.equal(new BN(1));
+        let etherBalanceAfter = await web3.eth.getBalance(acc5);
+        let etherBalance = new BN(etherBalanceAfter).add(ether('1'));
+        expect(Number(etherBalance)).to.be.closeTo(Number(etherBalanceBefore), Number(ether('0.0003')));
+        expectEvent(tx, "Buy", { buyer: acc5, ethers: ether('1') });
+      });
+
+    });
+    describe("False", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: Should fail if

File: test/MyNFT.test.js [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+      it("Error : all NFT's was sold", async () => {
+        let balanceBeforeNFT = await instanceNFT.balanceOf(acc6);
+        expect(balanceBeforeNFT).to.be.bignumber.equal(new BN(0));
+        await expectRevert(instanceNFT.buy({ from: acc6, value: ether('1') }), "Error : all NFT's was sold");
+        let balanceAfterNFT = await instanceNFT.balanceOf(acc6);
+        expect(balanceAfterNFT).to.be.bignumber.equal(new BN(0));
+      });
+
+      it("Error : payment value too low", async () => {
```
Comment (BhHaipls) [inline test/MyNFT.test.js]: or big, and add check if user send more funds, should fail also

**Review by BhHaipls — CHANGES_REQUESTED** [review]

File: contracts/MyNFT.sol [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+        require(msg.value == nftPrice, "Error : payment value too low");
+        require(
+            tokenCountId.current() <= tokenSupply,
+            "Error : all NFT's was sold"
+        );
+        _safeMint(msg.sender, tokenCountId.current());
+        emit Buy(msg.sender, tokenCountId.current(), msg.value);
+        tokenCountId.increment();
+    }
+
```
Comment (BhHaipls) [inline contracts/MyNFT.sol:85]: after sell , some eth will lock on contract, should add method for get paids user eth. Can be callo nly from Owner

**Review by BhHaipls — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Review by BhHaipls — APPROVED** [review]

**Review by BhHaipls — APPROVED** [review]
Approved

**Review by BhHaipls — CHANGES_REQUESTED** [review]
should change name of tests

**Review by BhHaipls — APPROVED** [review]
not need, completed, approved

**Pull request merged**